encoding/ssz/query/analyzer.go

```
@@ -0,0 +1,235 @@
+
+	case reflect.Struct:
+		return analyzeContainerType(typ)
+
+	case reflect.Ptr:
+		// Dereference pointer types.
+		return analyzeType(typ.Elem(), tag)
+
+	default:
+		return nil, fmt.Errorf("unsupported type for SSZ calculation: %v", typ.Kind())
+	}
+}
+
+// analyzeBasicType analyzes SSZ basic types (uintN, bool) and returns its info.
+func analyzeBasicType(typ reflect.Type) (*sszInfo, error) {
+	sszInfo := &sszInfo{
+		typ: typ,
+
+		// Every basic type is fixed-size and not variable.
+		isVariable: false,
+	}
+
+	switch typ.Kind() {
+	case reflect.Uint64:
+		sszInfo.sszType = UintN
+		sszInfo.fixedSize = 8
+	case reflect.Uint32:
+		sszInfo.sszType = UintN
+		sszInfo.fixedSize = 4
+	case reflect.Uint16:
+		sszInfo.sszType = UintN
+		sszInfo.fixedSize = 2
+	case reflect.Uint8:
+		sszInfo.sszType = UintN
+		sszInfo.fixedSize = 1
+	case reflect.Bool:
+		sszInfo.sszType = Boolean
+		sszInfo.fixedSize = 1
+	default:
+		return nil, fmt.Errorf("unsupported basic type for SSZ calculation: %v", typ.Kind())
+	}
+
+	return sszInfo, nil
+}
+
+// analyzeHomogeneousColType analyzes homogeneous collection types (e.g., List, Vector, Bitlist, Bitvector) and returns its SSZ info.
+// TODO: Support multi-dimensional parsing for tags. It only takes the first dimension into account.
+func analyzeHomogeneousColType(typ reflect.Type, tag *reflect.StructTag) (*sszInfo, error) {
+	if typ.Kind() != reflect.Slice {
+		return nil, fmt.Errorf("can only analyze slice types, got %v", typ.Kind())
+	}
+
+	if tag == nil {
+		return nil, fmt.Errorf("tag is required for slice types")
+	}
+
+	elementInfo, err := analyzeType(typ.Elem(), nil)
+	if err != nil {
+		return nil, fmt.Errorf("analyze element type for homogeneous collection: %w", err)
+	}
+
+	// 1. Check if the type is List/Bitlist by checking `ssz-max` tag.
+	sszMax := tag.Get(sszMaxTag)
+	if sszMax != "" {
+		dims := strings.Split(sszMax, ",")
+		limit, err := strconv.ParseUint(dims[0], 10, 64)
```

Since we only support one-dimensional lists, wouldn't it be better to return an error when `len(dims) > 1`?

---

Nice!